Documents dual authz with Mattermost (#6372)
gravitational/teleport: docs/pages/access-controls/guides/dual-authz.mdx

---
title: Dual Authorization
description: Dual Authorization for SSH and Kubernetes
---

# Dual Authorization

> You can set up Teleport to require require approvals of multiple members to perform some critical actions.

Reviewer: ```suggestion
You can set up Teleport to require the approval of multiple team members to perform some critical actions.
```

Here are the most common scenarios:

> - Improve security of your system and prevent one successful phishing attack to compromise your system.

Reviewer 2: "from compromising your system" sounds a bit better.

<Tabs>
<TabItem label="Mattermost">

Enable bot account creation in `System Console -> Integrations`.
Toggle `Enable Bot Account Creation`.

![Enable bots](../../../img/access-controls/dual-authz/mattermost-0-enable.png)

Go back to your team settings, navigate to `Integrations -> Bot Accounts`. Press `Add Bot Account`.

![Enable bots](../../../img/access-controls/dual-authz/mattermost-1-bot.png)

Add `Post All` permission on the new account.

![Enable bots](../../../img/access-controls/dual-authz/mattermost-2-all-permissions@2x.png)

Create the bot and save the access token.

</TabItem>
</Tabs>

Create a non-interactive bot `access-plugin` user and role.

```yaml
kind: user
metadata:
  name: access-plugin
spec:
  roles: ['access-plugin']
version: v2
---
kind: role
version: v3
metadata:
  name: access-plugin
spec:
  allow:
    rules:
      - resources: ['access_request']
        verbs: ['list', 'read']
      - resources: ['access_plugin_data']
        verbs: ['update']
    # teleport currently refuses to issue certs for a user with 0 logins,
    # this restriction may be lifted in future versions.
    logins: ['access-plugin-not-used']
```

<Admonition
  type="note"

>   title="Audit"

Reviewer 2: I think Title might be incorrect, maybe "Creating resources"?

>
  Here and below follow along and create yaml resources using `tctl create -f`:
  ```bash
  $ tctl create -f access.yaml
  ```
</Admonition>

**Export access-plugin cert**

> Teleport Plugin uses the access-plugin role and user to perform the approval.

Reviewer: ```suggestion
Teleport Plugin uses the `access-plugin` role and user to perform the approval.
```

We export the identify files, using tctl auth sign.

```bash
$ tctl auth sign --format=tls --user=access-plugin --out=auth --ttl=720h
```

> Teleport will generate `auth.crt`, `auth.key`, and `auth.cas` - a certificate, a private key, and a CA certs respectively.

Reviewer: ```suggestion
Teleport will generate `auth.crt`, `auth.key`, and `auth.cas` - a certificate, a private key, and a set of CA certs respectively.
```

**Install the plugin**

<Tabs>
<TabItem label="Download">
  ```bash
  $ curl -L https://get.gravitational.com/teleport-mattermost-v(=teleport.version=)-linux-amd64-bin.tar.gz
  $ tar -xzf teleport-mattermost-v(=teleport.version=)-linux-amd64-bin.tar.gz
  $ cd teleport-mattermost
  $ ./install
  ```
</TabItem>
<TabItem label="From Source">
  To install from source you need `git` and `go >= (=teleport.golang=)` installed.

  ```bash
  # Checkout teleport-plugins

>   $ git clone git@github.com:gravitational/teleport-plugins.git

Reviewer: ```suggestion
  $ git clone https://github.com/gravitational/teleport-plugins.git
```

  $ cd teleport-plugins/access/mattermost
  $ make
  ```
</TabItem>
</Tabs>

```bash
$ teleport-mattermost configure > /etc/teleport-mattermost.toml
```

Update the config with Teleport address, Mattermost URL and a bot token.

```toml
# example Mattermost configuration TOML file
[teleport]

> # For teleport cloud you can point to the proxy: "example.teleport.sh:443"

Reviewer: ```suggestion
# For Teleport Cloud you can point to the proxy: "example.teleport.sh:443"
```

  ```
</TabItem>
</Tabs>

Chatbot will notify both Alice and Ivan:

![Mattermost-Request](../../../img/access-controls/dual-authz/mattermost-5-request.png)

Alice and Ivan can review and approve request using Web UI or CLI:

<Tabs>
<TabItem label="Web UI">
![Teleport-Approve](../../../img/access-controls/dual-authz/teleport-6-ivan-approve.png)
</TabItem>
<TabItem label="CLI">
  ```bash
  $ tsh request list
  ID                                   User             Roles   Created (UTC)       Status
  ------------------------------------ ---------------  ------- ------------------- ------- 
  9c721e54-b049-4ef8-a7f6-c777aa066764 bob@example.com  dbadmin 03 Apr 21 03:58 UTC PENDING

  $tsh request review --approve --reason="hello" 9c721e54-b049-4ef8-a7f6-c777aa066764
  Successfully submitted review.  Request state: APPROVED
  ```
</TabItem>
</Tabs>

If Bob has created a request using CLI, he will assume it once it has been approved.
Bob can also assume granted access request roles using Web UI:

![Teleport Assume](../../../img/access-controls/dual-authz/teleport-7-bob-assume.png)

## Troubleshooting

**Cert errors**

> You may be getting certificate errors if Teleport's auth server missing

Reviewer 2: auth server **is** missing address
